server/lib/filter.js

```diff
@@ -195,12 +195,23 @@ class FilterParser {
   applyQuery(sql) {
     // optionally call utility method to parse all remaining options as simple
     // equality filters into `_statements`
-    let limitCondition = this._parseLimit();
+    const limitCondition = this._parseLimit();
     this._parseDefaultFilters();
-    let conditionStatements = this._parseStatements();
-    let order = this._order;
+    const conditionStatements = this._parseStatements();
+    const order = this._order;
+    let statement;
+
+    // allows a developer to specify the position of the WHERE injection
+    // by default, it will simply concat
+    const hasWhereInjection = sql.includes('{{ WHERE }}');
```

Comment:
This is an interesting solution! It took me a minute to figure out exactly what was going on here when looking at it from an API perspective.

I think this is because `FilterParser` didn't anticipate correctly ordering statements like `ORDER BY`, `GROUP BY`, `LIMIT` etc. and so is limited. 

The quick solution for `ORDER BY` was the method `setOrder()`, limit is done automagically and somewhat confusingly with `_parseLimit()` and now group by can be handled by adding `{{ WHERE }}` into the template. 

I suggest we design the API here so that it can handle each of these in a uniform way, I think there are a number of different solutions. 

1. Use the `setX` method for each eventuality, `setOrder`, `setGroup`, `setLimit` etc. These methods will accept the SQL that should be templated and enter them in the correct order. 

2. Use the templating method for each eventuality, in this way the programmer will specify the SQL with a `{{ WHERE }}` line and add limits, groups and order manually. 

3. Limit the scope of filter parser to only deal with the `WHERE` condition; this would then leave it up to the programmer how the SQL is templated. For example: 

```js
// equivalent of filters.applyQuery but not concerned with the sql itself
const where = filters.parseFilterConditions();
const whereParams = filters.parameters();

// most complex example
const query = `${sql} WHERE ${where} ${groupSql} ${orderSql} ${limitSql};

db.exec(query, whereParams) //...
```


- Do you think that it's worth ensuring that these APIs are uniform or is the current method okay? 
- Do you have any thoughts on the proposals here?